### PR TITLE
Add Option to Force Unwind Tables

### PR DESCRIPTION
**Pull request by lenary**
When panic != unwind, `nounwind` is added to all functions for a target.
This can cause issues when a panic happens with RUST_BACKTRACE=1, as
there needs to be a way to reconstruct the backtrace. There are three
possible sources of this information: forcing frame pointers (for which
an option exists already), debug info (for which an option exists), or
unwind tables.

Especially for embedded devices, forcing frame pointers can have code
size overheads (RISC-V sees ~10% overheads, ARM sees ~2-3% overheads).
In production code, it can be the case that debug info is not kept, so it is useful
to provide this third option, unwind tables, that users can use to
reconstruct the call stack. Reconstructing this stack is harder than
with frame pointers, but it is still possible.

---

This came up in discussion on #69890, and turned out to be a fairly simple addition.

r? @hanna-kruppe

### CODE REVIEW

File: src/doc/rustc/src/codegen-options/index.md
Comment (jonas-schievink) [inline src/doc/rustc/src/codegen-options/index.md]:
```suggestion
## force-unwind-tables
```

File: src/librustc_session/session.rs
Comment (hanna-kruppe) [inline src/librustc_session/session.rs]: As implemented, this allows users to turn off unwinding tables even if the target strictly requires them. That seems like a footgun.

**Comment by hanna-kruppe** [comment]
I'm obviously in favor of having this option in some way, but since `-C` options are effectively "stable" I believe @rust-lang/compiler should sign off the new option.

File: src/librustc_session/session.rs
Comment (oli-obk) [reply src/librustc_session/session.rs]: should we just error in case of such a conflict?

Comment (nagisa) [reply src/librustc_session/session.rs]: Emitting an error has prior art in rustc so, yes.

**Comment by nagisa** [comment]
@rfcbot merge

This introduces an insta-stable argument to control a fairly well understood behaviour of the compilers (not just rustc, but also clang/gcc (`-funwind-tables`) and others). While the PR itself has some kinks to iron out, this FCP is specifically to gather consensus on merging an insta-stable flag to control whether unwinding tables are emit.

**Comment by nagisa** [comment]
@rfcbot fcp merge

This introduces an insta-stable argument to control a fairly well understood behaviour of the compilers (not just rustc, but also clang/gcc (`-funwind-tables`) and others). While the PR itself has some kinks to iron out, this FCP is specifically to gather consensus on merging an insta-stable flag to control whether unwinding tables are emit.

File: src/librustc_session/session.rs
Comment (lenary) [reply src/librustc_session/session.rs]: Ok, Will do when I get back to my work computer.

**Comment by Amanieu** [comment]
As I commented in #69890:

> I personally think that this should be added as a third option to `-C panic`, called `abort-with-backtrace`. All it does is enable `uwtable` to guarantee that backtraces can be reliably generated. The reason for making it separate from `-C panic=abort` is that this comes at a slight code size increase (due to the need to preserve CSRs even for `noreturn` functions) and a binary size increase from the `.eh_frame` section.

> On targets which do not support unwinding (e.g. thumb* & riscv*), the standard library would then be built with `-C panic=abort-with-backtrace` to ensure that reconstructing backtraces is always possible on these targets.

**Comment by hanna-kruppe** [comment]
@Amanieu Can you elaborate why you think a new `-C panic` value is better than the independent flag that is proposed here? I don't see any advantage, and some downsides:

- The panic=unwind/abort choice ties into several other things (which panic runtime is implicitly linked into leaf crates, which rlibs can be linked together) so adding another panic strategy brings unnecessary questions/baggage along.
- While the motivation for this feature is backtraces on embedded targets, it's not inherently specific to panic=abort. `-C force-unwind-tables=on` could also help with generating backtraces *outside of panics* (e.g., profiling or error logging), since some functions can't unwind and wouldn't get an `.eh_frame` entry even with panic=unwind.

File: src/librustc_session/session.rs
Comment (lenary) [reply src/librustc_session/session.rs]: I believe that I've done this in the right place: `validate_commandline_args_with_session_available`. This check only takes into account the target definition, not the `no_landing_pads` that is checked at attribute insertion time.

**Comment by Amanieu** [comment]
One advantage of `-C panic` is that it can be [specified in Cargo.toml](https://doc.rust-lang.org/edition-guide/rust-2018/error-handling-and-panics/aborting-on-panic.html).

> -C force-unwind-tables=on could also help with generating backtraces outside of panics (e.g., profiling or error logging), since some functions can't unwind and wouldn't get an .eh_frame entry even with panic=unwind.

AFAIK that's not true, when compiling with `panic=unwind` all functions, regardless of whether they unwind or not, will get the `uwtable` attribute. The issue with missing backtraces only occurs with `panic=abort`, hence why I suggested making this an extension of `panic=abort`.

**Comment by hanna-kruppe** [comment]
> One advantage of `-C panic` is that it can be [specified in Cargo.toml](https://doc.rust-lang.org/edition-guide/rust-2018/error-handling-and-panics/aborting-on-panic.html).

Cargo [has its own validation of this Cargo.toml entry](https://github.com/rust-lang/cargo/blob/dcddf15203b91b1769060e38921974cddce4c408/src/cargo/util/toml/mod.rs#L517), so some Cargo changes are required to enable this. And if Cargo changes are required anyway, I don't think it would be significantly harder to add a new `force-unwind-tables` Cargo.toml key for this.  

> > -C force-unwind-tables=on could also help with generating backtraces outside of panics (e.g., profiling or error logging), since some functions can't unwind and wouldn't get an .eh_frame entry even with panic=unwind.
> 
> AFAIK that's not true, when compiling with `panic=unwind` all functions, regardless of whether they unwind or not, will get the `uwtable` attribute. The issue with missing backtraces only occurs with `panic=abort`, hence why I suggested making this an extension of `panic=abort`.

Ah, right, this is indeed mostly the current behavior. But it is not true [for allocator shims](https://github.com/rust-lang/rust/blob/c20d7eecbc0928b57da8fe30b2ef8528e2bdd5be/src/librustc_codegen_llvm/allocator.rs#L57), and I don't think the current behavior for other functions is *intended* or can be expected to last forever.

For historic reasons, certain code paths in the compiler currently decide some unwinding-related things based on the global `no_landing_pads` property instead of a more precise per-function predicate that e.g. takes `#[unwind]` attributes into account. Once we clean up our act around unwinding (see also: #65303), I expect that the logic for `uwtable` being added will also be based on "can this function unwind / are we adding `nounwind` to it". Generally, the incentives for dropping `uwtable` from e.g. `#[unwind(aborts)]` functions are the same as dropping `uwtable` from *all* functions in panic=abort mode, just proportionally smaller.

**Comment by Amanieu** [comment]
Fair enough, I'll withdraw my objection to making this a `-C` flag.

**Comment by nikomatsakis** [comment]
@rfcbot reviewed

**Comment by Zoxc** [comment]
Should we avoid emitting unwind tables by default on Windows (and let people opt-in with this flag)? If not, we probably want an option to turn off unwind tables which this doesn't enable.

**Comment by hanna-kruppe** [comment]
Given the many different mechanisms and tools that rely on SEH, and [Windows gurus stating](https://stackoverflow.com/a/45333553) that pretty much *all* functions should have unwinding information on x64 Windows, it does not seem very wise to ever omit unwinding tables. Even an opt-out seems dubious to me, and as far as I can tell it's not a feature anyone has ever requested. In fact, after poking at it for a little bit, I can't find any such opt out in MSVC, even when compiling plain C code.

In any case, should we ever change our mind about this, we can still change `-Cforce-unwind-tables=no` from reporting an error if the target has them enabled by default to override the target's default instead. This is backwards compatible, so it can (and should) be a separate discussion at any point after the flag lands.

**Comment by lenary** [comment]
My understanding is that the `self.no_landing_pads()` will always be false on windows (that code used to have windows-specific comments by it, which were out of date), so this patch should ensure that you cannot remove unwind tables on windows. 

I don't know if i should include that check in the error reporting here: https://github.com/rust-lang/rust/pull/69984/files#diff-eadaa1c9e135801995ae29eaec608049R1173 - happy to update the patch to do so if needed, not sure how it will affect testing.

**Comment by hanna-kruppe** [comment]
> My understanding is that the `self.no_landing_pads()` will always be false on windows (that code used to have windows-specific comments by it, which were out of date)

I'm pretty sure this is wrong. `Session::no_landing_pads()` (while it still exists; #70175 removes it) only looks at `-C panic` and `-Z no-landing-pads`, neither of which has Windows-related restrictions, so it can easily return true on Windows targets. The target spec field `requires_uwtable` is taken into account independently of what `no_landing_pads()` says. [This PR even refactors that condition](https://github.com/rust-lang/rust/pull/69984/files#diff-4818f50800787c101a97fba3eae6c56dL296-R280) to introduce the `must_emit_uwtable`.

However, your conclusion (that users can't remove unwinding information on Windows targets) is correct as far as I can tell.

File: src/librustc_session/session.rs
Comment (jonas-schievink) [inline src/librustc_session/session.rs:1175]:
```suggestion
                "target requires unwind tables, they cannot be disabled with \
```

(to match other error messages)

Also, do we want a test for this?

**Review by jonas-schievink — CHANGES_REQUESTED** [review]

**Comment by nnethercote** [comment]
When you rebase over #71374 please make sure you preserve the new alphabetical order among the options, in the code, tests, and documentation. Thanks!

**Comment by crlf0710** [comment]
@lenary Ping from triage, would you mind resolving the merge conflicts and address the comments above, so we can proceed with this?

**Comment by lenary** [comment]
Will resolve conflicts (and documentation) today.

File: src/librustc_session/session.rs
Comment (lenary) [reply src/librustc_session/session.rs]: I'm not sure how to write a test that works in a way that it can be run (and pass) on multiple targets.

**Comment by lenary** [comment]
Ok, I've pushed an update which addresses all the outstanding issues, I believe.

File: src/librustc_session/session.rs
Comment (hanna-kruppe) [reply src/librustc_session/session.rs]: Tests can be restricted to specific targets, in this case I think a compile-fail test with `// only-x86_64-windows-msvc` header should suffice.

File: src/doc/rustc/src/codegen-options/index.md
Comment (hanna-kruppe) [inline src/doc/rustc/src/codegen-options/index.md:108]: I think this sentence is no longer accurate, an error is generated instead.

**Review by hanna-kruppe — CHANGES_REQUESTED** [review]
I'd like to have the test @jonas-schievink suggested, and found one more nit. r=me after that.

**Comment by hanna-kruppe** [comment]
Thanks! 

@bors r+

File: src/doc/rustc/src/codegen-options/index.md
Comment (lenary) [reply src/doc/rustc/src/codegen-options/index.md]: Updated!

File: src/librustc_session/session.rs
Comment (lenary) [reply src/librustc_session/session.rs]: Let's hope I got this line right!

**Comment by hanna-kruppe** [comment]
Seems spurious, @bors retry

**Pull request merged**
